src/keyring/KeyringController.test.ts

```diff
@@ -489,4 +500,344 @@ describe('KeyringController', () => {
+      {
+        setAccountLabel: preferences.setAccountLabel.bind(preferences),
+        removeIdentity: preferences.removeIdentity.bind(preferences),
+        syncIdentities: preferences.syncIdentities.bind(preferences),
+        updateIdentities: preferences.updateIdentities.bind(preferences),
+        setSelectedAddress: preferences.setSelectedAddress.bind(preferences),
+      },
+      baseConfig,
+    );
+
+    signProcessKeyringController.createNewVaultAndKeychain(password);
+
+    it('should add a new QR keyring when first call QR keyring related methods', async () => {
+      expect(
+        signProcessKeyringController.state.keyrings.find(
+          (keyring) => keyring.type === KeyringTypes.qr,
+        ),
+      ).toBeUndefined();
+      await signProcessKeyringController.getQRKeyringState();
+      expect(
+        signProcessKeyringController.state.keyrings.find(
+          (keyring) => keyring.type === KeyringTypes.qr,
+        ),
+      ).toBeDefined();
+    });
+
+    it('should setup QR keyring with crypto-hdkey', async () => {
+      setTimeout(
+        () =>
+          signProcessKeyringController.submitQRCryptoHDKey(
+            'a902f40358210219218eb65839d08bde4338640b03fdbbdec439ef880d397c2f881282c5b5d135045820e65ed63f52e3e93d48ffb55cd68c6721e58ead9b29b784b8aba58354f4a3d92905d90131a201183c020006d90130a30186182cf5183cf500f5021a5271c071030307d90130a2018400f480f40300081a625f3e6209684b657973746f6e650a706163636f756e742e7374616e64617264',
+          ),
+        100,
+      );
+      const firstPage = await signProcessKeyringController.connectQRHardware(0);
+      expect(firstPage).toHaveLength(5);
+      expect(firstPage[0].index).toBe(0);
+
+      const secondPage = await signProcessKeyringController.connectQRHardware(
+        1,
+      );
+      expect(secondPage).toHaveLength(5);
+      expect(secondPage[0].index).toBe(5);
+
+      const goBackPage = await signProcessKeyringController.connectQRHardware(
+        -1,
+      );
+      expect(goBackPage).toStrictEqual(firstPage);
+
+      await signProcessKeyringController.unlockQRHardwareWalletAccount(0);
+      await signProcessKeyringController.unlockQRHardwareWalletAccount(1);
+      await signProcessKeyringController.unlockQRHardwareWalletAccount(2);
+
+      const qrKeyring = signProcessKeyringController.state.keyrings.find(
+        (keyring) => keyring.type === KeyringTypes.qr,
+      );
+      expect(qrKeyring?.accounts).toHaveLength(3);
+    });
+
+    const composeMockSignature = (requestId: string, signature: string) => {
+      const rlpSignatureData = Buffer.from(signature, 'hex');
+      const idBuffer = uuid.parse(requestId) as Uint8Array;
+      const ethSignature = new ETHSignature(
+        rlpSignatureData,
+        Buffer.from(idBuffer),
+      );
+      return ethSignature.toCBOR().toString('hex');
+    };
+
+    it('should sign message with QR keyring', async () => {
+      setTimeout(async () => {
```

Comment:
That's because QR keyring needs some user interaction to perform transaction signing and wallet syncing. So the implementation used an event emitter to wait for the QR code from Keystone side.

Reply:
Hmm, I see. I'm wary of adding timeouts because I feel like they could lead to intermittently failing tests. This code will also run after the test is finished running, and if any part of it fails, it won't cause the test to fail (and Jest will give us a warning). Is there any way to get around this?

Reply:
Hi @mcmire , thanks for your suggestion and I just stub some implementation with QR Keyring to write the unit tests so that we do not need setTimeout.

Reply:
Thank you so much! That's much better :)